File: packages/telemetry/python/tests/telemetry/span_filter_test.py
"""Tests for smart span export filtering."""

from unittest.mock import Mock

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter

from latitude_telemetry.constants import SCOPE_LATITUDE
from latitude_telemetry.telemetry.span_filter import (
    ExportFilterSpanProcessor,
    build_should_export_span,
    is_default_export_span,
    is_gen_ai_or_llm_attribute_span,
    is_latitude_instrumentation_span,
)


def _readable_span(*, scope_name: str = "", attributes: dict | None = None):
    span = Mock()
    scope = Mock()
    scope.name = scope_name
    span.instrumentation_scope = scope
    span.attributes = attributes or {}
    return span


class TestIsLatitudeInstrumentationSpan:
    def test_matches_nested_scope(self) -> None:
        assert is_latitude_instrumentation_span(_readable_span(scope_name=f"{SCOPE_LATITUDE}.manual"))

    def test_rejects_foreign_scope(self) -> None:
        assert not is_latitude_instrumentation_span(_readable_span(scope_name="express"))


class TestIsGenAiOrLlmAttributeSpan:
    def test_gen_ai_prefix(self) -> None:
        assert is_gen_ai_or_llm_attribute_span(
            _readable_span(attributes={"gen_ai.request.model": "gpt-4"}),
        )

    def test_http_only(self) -> None:
        assert not is_gen_ai_or_llm_attribute_span(
            _readable_span(attributes={"http.method": "GET"}),
        )


class TestIsDefaultExportSpan:
    def test_rejects_http_instrumentation(self) -> None:
        assert not is_default_export_span(
            _readable_span(
                scope_name="opentelemetry.instrumentation.requests",
                attributes={"http.method": "GET"},
            ),
        )

    def test_accepts_openai_instrumentation(self) -> None:
        assert is_default_export_span(_readable_span(scope_name="opentelemetry.instrumentation.openai"))

    def test_accepts_openinference_instrumentation(self) -> None:
        assert is_default_export_span(_readable_span(scope_name="openinference.instrumentation.langchain"))

    def test_accepts_traceloop_substring(self) -> None:
        assert is_default_export_span(_readable_span(scope_name="traceloop.instrumentation.openai"))

    def test_accepts_langsmith_substring(self) -> None:
        assert is_default_export_span(_readable_span(scope_name="my.langsmith.tracer"))

    def test_accepts_litellm_substring(self) -> None:
        assert is_default_export_span(_readable_span(scope_name="litellm.proxy"))


class TestBuildShouldExportSpan:
    def test_disable_smart_filter(self) -> None:
        pred = build_should_export_span(disable_smart_filter=True)
        assert pred(_readable_span(scope_name="opentelemetry.instrumentation.requests"))

    def test_blocked_scopes(self) -> None:
        pred = build_should_export_span(
            blocked_instrumentation_scopes=["opentelemetry.instrumentation.openai"],
        )
        assert not pred(_readable_span(scope_name="opentelemetry.instrumentation.openai"))
        assert pred(_readable_span(scope_name="opentelemetry.instrumentation.anthropic"))

    def test_should_export_span_extra(self) -> None:
        pred = build_should_export_span(
            should_export_span=lambda s: s.instrumentation_scope.name == "custom.scope",
        )
        assert pred(_readable_span(scope_name="custom.scope"))
        assert not pred(_readable_span(scope_name="express"))


class TestExportFilterParentChainPromotion:
    def setup_method(self) -> None:
        self.exporter = InMemorySpanExporter()
        self.provider = TracerProvider()
        self.provider.add_span_processor(
            ExportFilterSpanProcessor(
                build_should_export_span(),
                SimpleSpanProcessor(self.exporter),
            )
        )

    def teardown_method(self) -> None:
        self.provider.shutdown()
        self.exporter.clear()

    def test_exports_unstamped_parent_when_kept_child_ends_first(self) -> None:
        tracer = self.provider.get_tracer("opentelemetry.instrumentation.net")
        parent = tracer.start_span("tcp.connect.parent")
        child = tracer.start_span(
            "tcp.connect",
            context=trace.set_span_in_context(parent),
            attributes={"latitude.tags": '["flagger"]'},
        )
        child.end()
        parent.end()
        self.provider.force_flush()

        names = sorted(span.name for span in self.exporter.get_finished_spans())
        assert names == ["tcp.connect", "tcp.connect.parent"]

    def test_flushes_already_dropped_parent_when_child_kept(self) -> None:
        tracer = self.provider.get_tracer("opentelemetry.instrumentation.net")
        parent = tracer.start_span("http.request")
        child_ctx = trace.set_span_in_context(parent)
        parent.end()

        child = tracer.start_span(
            "tcp.connect",
            context=child_ctx,
            attributes={"latitude.capture.name": "flagger.draft"},
        )
        child.end()
        self.provider.force_flush()

        names = sorted(span.name for span in self.exporter.get_finished_spans())
        assert names == ["http.request", "tcp.connect"]

    def test_still_drops_spans_with_no_kept_descendant(self) -> None:
        tracer = self.provider.get_tracer("opentelemetry.instrumentation.net")
        parent = tracer.start_span("dns.lookup")
        child = tracer.start_span("tcp.connect", context=trace.set_span_in_context(parent))
        child.end()
        parent.end()
        self.provider.force_flush()

        assert self.exporter.get_finished_spans() == ()

    def test_promotes_multi_level_ancestor_chain(self) -> None:
        tracer = self.provider.get_tracer("opentelemetry.instrumentation.net")
        root = tracer.start_span("http.request")
        mid = tracer.start_span("tls.connect", context=trace.set_span_in_context(root))
        leaf = tracer.start_span(
            "tcp.connect",
            context=trace.set_span_in_context(mid),
            attributes={"gen_ai.request.model": "gpt-4"},
        )
        leaf.end()
        mid.end()
        root.end()
        self.provider.force_flush()

        names = sorted(span.name for span in self.exporter.get_finished_spans())
        assert names == ["http.request", "tcp.connect", "tls.connect"]

File: packages/telemetry/typescript/src/sdk/span-filter.test.ts
import { context, trace } from "@opentelemetry/api"
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks"
import {
  InMemorySpanExporter,
  NodeTracerProvider,
  type ReadableSpan,
  SimpleSpanProcessor,
} from "@opentelemetry/sdk-trace-node"
import { afterAll, beforeAll, beforeEach, describe, expect, it } from "vitest"
import { SCOPE_LATITUDE } from "../constants/scope.ts"
import {
  buildShouldExportSpan,
  ExportFilterSpanProcessor,
  isDefaultExportSpan,
  isGenAiOrLlmAttributeSpan,
  isLatitudeInstrumentationSpan,
} from "./span-filter.ts"

function mockSpan(overrides: { scopeName?: string; attributes?: Record<string, unknown> }): ReadableSpan {
  const { scopeName = "", attributes = {} } = overrides
  return {
    instrumentationScope: { name: scopeName },
    attributes,
  } as unknown as ReadableSpan
}

describe("isLatitudeInstrumentationSpan", () => {
  it("matches manual and nested latitude scopes", () => {
    expect(isLatitudeInstrumentationSpan(mockSpan({ scopeName: `${SCOPE_LATITUDE}.manual` }))).toBe(true)
    expect(isLatitudeInstrumentationSpan(mockSpan({ scopeName: SCOPE_LATITUDE }))).toBe(true)
    expect(isLatitudeInstrumentationSpan(mockSpan({ scopeName: "express" }))).toBe(false)
  })
})

describe("isGenAiOrLlmAttributeSpan", () => {
  it("matches gen_ai and llm attribute prefixes", () => {
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "gen_ai.request.model": "gpt-4" } }))).toBe(true)
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "llm.model_name": "x" } }))).toBe(true)
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "openinference.span.kind": "CHAIN" } }))).toBe(true)
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "eve.session.id": "sess-1" } }))).toBe(true)
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "flue.run.id": "run-1" } }))).toBe(true)
    expect(isGenAiOrLlmAttributeSpan(mockSpan({ attributes: { "http.route": "/api" } }))).toBe(false)
  })
})

describe("Vercel AI SDK v7 spans", () => {
  // v7's `@ai-sdk/otel` `OpenTelemetry` integration emits GenAI SemConv spans on the
  // `gen_ai` tracer scope (operation names invoke_agent / chat / execute_tool).
  it("exports v7 GenAI spans (gen_ai.* attributes)", () => {
    for (const op of ["invoke_agent", "chat", "execute_tool", "embeddings", "rerank"]) {
      const span = mockSpan({ scopeName: "gen_ai", attributes: { "gen_ai.operation.name": op } })
      expect(isGenAiOrLlmAttributeSpan(span)).toBe(true)
      expect(isDefaultExportSpan(span)).toBe(true)
      expect(buildShouldExportSpan({})(span)).toBe(true)
    }
  })

  // v7's `LegacyOpenTelemetry` (and v6 built-in) emit `ai.*` spans on the `ai` scope.
  it("exports v7/v6 legacy spans (ai.* attributes)", () => {
    const span = mockSpan({ scopeName: "ai", attributes: { "ai.operationId": "ai.streamText.doStream" } })
    expect(isGenAiOrLlmAttributeSpan(span)).toBe(true)
    expect(buildShouldExportSpan({})(span)).toBe(true)
  })

  it("still drops non-LLM spans regardless of scope", () => {
    const span = mockSpan({ scopeName: "gen_ai", attributes: { "http.route": "/api" } })
    expect(buildShouldExportSpan({})(span)).toBe(false)
  })
})

describe("isDefaultExportSpan", () => {
  it("rejects generic HTTP instrumentation", () => {
    expect(
      isDefaultExportSpan(
        mockSpan({
          scopeName: "opentelemetry.instrumentation.requests",
          attributes: { "http.method": "GET" },
        }),
      ),
    ).toBe(false)
  })

  it("accepts known LLM OTel scopes", () => {
    expect(isDefaultExportSpan(mockSpan({ scopeName: "opentelemetry.instrumentation.openai" }))).toBe(true)
    expect(isDefaultExportSpan(mockSpan({ scopeName: "openinference.instrumentation.langchain" }))).toBe(true)
  })

  it("accepts traceloop substring scopes", () => {
    expect(isDefaultExportSpan(mockSpan({ scopeName: "traceloop.instrumentation.openai" }))).toBe(true)
  })

  it("accepts langsmith substring scopes", () => {
    expect(isDefaultExportSpan(mockSpan({ scopeName: "my.langsmith.tracer" }))).toBe(true)
  })

  it("accepts litellm substring scopes", () => {
    expect(isDefaultExportSpan(mockSpan({ scopeName: "litellm.proxy" }))).toBe(true)
  })
})

describe("buildShouldExportSpan", () => {
  it("exports everything when smart filter disabled", () => {
    const pred = buildShouldExportSpan({ disableSmartFilter: true })
    expect(pred(mockSpan({ scopeName: "opentelemetry.instrumentation.requests" }))).toBe(true)
  })

  it("respects blocked scopes", () => {
    const pred = buildShouldExportSpan({
      blockedInstrumentationScopes: ["opentelemetry.instrumentation.openai"],
    })
    expect(pred(mockSpan({ scopeName: "opentelemetry.instrumentation.openai" }))).toBe(false)
    expect(pred(mockSpan({ scopeName: "opentelemetry.instrumentation.anthropic" }))).toBe(true)
  })

  it("composes with shouldExportSpan", () => {
    const pred = buildShouldExportSpan({
      shouldExportSpan: (s) => s.instrumentationScope?.name === "my.custom.scope",
    })
    expect(pred(mockSpan({ scopeName: "my.custom.scope" }))).toBe(true)
    expect(pred(mockSpan({ scopeName: "express" }))).toBe(false)
  })
})

describe("ExportFilterSpanProcessor parent-chain promotion", () => {
  const exporter = new InMemorySpanExporter()
  let provider: NodeTracerProvider
  let previousProvider: ReturnType<typeof trace.getTracerProvider>

  beforeAll(() => {
    context.setGlobalContextManager(new AsyncLocalStorageContextManager())
    previousProvider = trace.getTracerProvider()
    provider = new NodeTracerProvider({
      spanProcessors: [new ExportFilterSpanProcessor(buildShouldExportSpan({}), new SimpleSpanProcessor(exporter))],
    })
    trace.setGlobalTracerProvider(provider)
  })

  beforeEach(() => {
    exporter.reset()
  })

  afterAll(async () => {
    trace.setGlobalTracerProvider(previousProvider)
    await provider.shutdown()
  })

  it("exports an unstamped parent when a kept child ends first", async () => {
    const net = trace.getTracer("opentelemetry.instrumentation.net")
    const parent = net.startSpan("tcp.connect.parent")
    const child = net.startSpan(
      "tcp.connect",
      { attributes: { "latitude.tags": '["flagger"]' } },
      trace.setSpan(context.active(), parent),
    )
    child.end()
    parent.end()
    await provider.forceFlush()

    const names = exporter
      .getFinishedSpans()
      .map((s) => s.name)
      .sort()
    expect(names).toEqual(["tcp.connect", "tcp.connect.parent"])
    const exportedChild = exporter.getFinishedSpans().find((s) => s.name === "tcp.connect")
    const exportedParent = exporter.getFinishedSpans().find((s) => s.name === "tcp.connect.parent")
    expect(exportedChild?.parentSpanContext?.spanId).toBe(exportedParent?.spanContext().spanId)
  })

  it("flushes an already-dropped parent when a later child is kept", async () => {
    const net = trace.getTracer("opentelemetry.instrumentation.net")
    const parent = net.startSpan("http.request")
    const childCtx = trace.setSpan(context.active(), parent)
    // End parent before the kept child — unusual but possible with async instrumentation.
    parent.end()

    const child = net.startSpan("tcp.connect", { attributes: { "latitude.capture.name": "flagger.draft" } }, childCtx)
    child.end()
    await provider.forceFlush()

    const names = exporter
      .getFinishedSpans()
      .map((s) => s.name)
      .sort()
    expect(names).toEqual(["http.request", "tcp.connect"])
  })

  it("still drops spans with no kept descendant", async () => {
    const net = trace.getTracer("opentelemetry.instrumentation.net")
    const parent = net.startSpan("dns.lookup")
    const child = net.startSpan("tcp.connect", undefined, trace.setSpan(context.active(), parent))
    child.end()
    parent.end()
    await provider.forceFlush()

    expect(exporter.getFinishedSpans()).toHaveLength(0)
  })

  it("promotes a multi-level ancestor chain", async () => {
    const net = trace.getTracer("opentelemetry.instrumentation.net")
    const root = net.startSpan("http.request")
    const mid = net.startSpan("tls.connect", undefined, trace.setSpan(context.active(), root))
    const leaf = net.startSpan(
      "tcp.connect",
      { attributes: { "gen_ai.request.model": "gpt-4" } },
      trace.setSpan(context.active(), mid),
    )
    leaf.end()
    mid.end()
    root.end()
    await provider.forceFlush()

    const names = exporter
      .getFinishedSpans()
      .map((s) => s.name)
      .sort()
    expect(names).toEqual(["http.request", "tcp.connect", "tls.connect"])
  })

  it("still promotes ancestors after the dropped-span buffer fills", async () => {
    const net = trace.getTracer("opentelemetry.instrumentation.net")
    const parent = net.startSpan("http.request")
    parent.end()

    for (let i = 0; i < 2048; i++) {
      net.startSpan(`noise-${i}`).end()
    }

    const child = net.startSpan(
      "tcp.connect",
      { attributes: { "gen_ai.request.model": "gpt-4" } },
      trace.setSpan(context.active(), parent),
    )
    child.end()
    await provider.forceFlush()

    const names = exporter
      .getFinishedSpans()
      .map((s) => s.name)
      .sort()
    expect(names).toEqual(["http.request", "tcp.connect"])
  })
})

describe("ExportFilterSpanProcessor blocked scopes", () => {
  it("does not export a blocked ancestor even when a kept child promotes the chain", async () => {
    const exporter = new InMemorySpanExporter()
    const provider = new NodeTracerProvider({
      spanProcessors: [
        new ExportFilterSpanProcessor(
          buildShouldExportSpan({
            blockedInstrumentationScopes: ["opentelemetry.instrumentation.net"],
          }),
          new SimpleSpanProcessor(exporter),
          { blockedInstrumentationScopes: ["opentelemetry.instrumentation.net"] },
        ),
      ],
    })

    const http = provider.getTracer("opentelemetry.instrumentation.http")
    const net = provider.getTracer("opentelemetry.instrumentation.net")
    const parent = http.startSpan("http.request")
    const blocked = net.startSpan("tcp.connect", undefined, trace.setSpan(context.active(), parent))
    const leaf = http.startSpan(
      "genai.child",
      { attributes: { "gen_ai.request.model": "gpt-4" } },
      trace.setSpan(context.active(), blocked),
    )
    leaf.end()
    blocked.end()
    parent.end()
    await provider.forceFlush()

    const names = exporter
      .getFinishedSpans()
      .map((s) => s.name)
      .sort()
    expect(names).toEqual(["genai.child", "http.request"])
    await provider.shutdown()
  })
})
